core/src/main/resources/error/error-classes.json

```diff
@@ -876,6 +876,13 @@
     ],
     "sqlState" : "42000"
   },
+  "NOT_ENOUGH_DATA_COLUMNS" : {
+    "message" : [
+      "Cannot write to <tableName>, not enough data columns:",
```

Comment:
The condition is more strict:
https://github.com/apache/spark/blob/a80899f8bef84bb471606f0bbef889aa5e81628b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L380

it checks exact number of columns but not `enough data columns:`.

Reply:
Can we re-use the error class `NUM_COLUMNS_MISMATCH` in the case? If not, would it possible to make it more generic to cover the case too.

Reply:
Let me try

Reply:
@MaxGekk Can I change the template of `NUM_COLUMNS_MISMATCH ` error message to "<operator> can only be performed on tables with the same number of columns, but the first table (<tableName>)'s columns(length: <refLength>): [refColumns], the (ordinal: <invalidOrdinalNum>)'s columns(length: <invalidLength>): [invalidColumns]." ?

The original is "<operator> can only be performed on tables with the same number of columns, but the first table has <refNumColumns> columns and the <invalidOrdinalNum> table has <invalidNumColumns> columns."

Reply:
Given that we have columns defaults, at least in the case of insert it's not really about the target table, but the column list.
I take it the same error is raised on UPDATE T SET  (...) = (...)?  and UNION/EXCEPT/...  ( ) .. IN ( SELECT ...) ?

How about <operator> expects matching number of columns. But the left side has <leftNumCols> while the right side has <rightNumCols>.

Alternatively we could talk about "assignments" using a different error class (?)

Reply:
I will open a new pr to make the error class NUM_COLUMNS_MISMATCH more generic.

Reply:
@panbingkun @srielau 
New Pr: https://github.com/apache/spark/pull/38937

After above, I will continue this pr.

Reply:
@panbingkun The PR https://github.com/apache/spark/pull/38937 has been merged already. Please, continue with this.

Reply:
OK, Let me continue to do it.